src/client_builder.rs

```diff
@@ -61,35 +129,88 @@ where
         ClientBuilder {
             domain,
             port,
-            starttls: false,
+            mode: ConnectionMode::Tcp,
+            tls_kind: TlsKind::Any,
+            no_tls_verify: false,
         }
     }
 
     /// Use [`STARTTLS`](https://tools.ietf.org/html/rfc2595) for this connection.
-    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
     pub fn starttls(&mut self) -> &mut Self {
-        self.starttls = true;
+        self.mode = ConnectionMode::StartTls;
         self
     }
 
-    /// Return a new [`Client`] using a `native-tls` transport.
-    #[cfg(feature = "native-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "native-tls")))]
-    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = TlsConnector::builder().build()?;
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    /// Use TLS for this connection.
+    pub fn tls(&mut self) -> &mut Self {
+        self.mode = ConnectionMode::Tls;
+        self
     }
 
-    /// Return a new [`Client`] using `rustls` transport.
-    #[cfg(feature = "rustls-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
-    pub fn rustls(&mut self) -> Result<Client<RustlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = RustlsConnector::new_with_native_certs()?;
-            Ok(ssl_conn.connect(domain, tcp)?)
-        })
+    /// Sets the TLS backend to use for this connection.
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
+        self.tls_kind = kind;
+        self
+    }
+
+    /// Controls the use of certificate validation.
+    ///
+    /// Defaults to `false`.
+    ///
+    /// # Warning
+    ///
+    /// You should only use this as a last resort
+    ///
+    /// See [`native_tls::TlsConnectorBuilder::danger_accept_invalid_certs`],
+    /// [`native_tls::TlsConnectorBuilder::danger_accept_invalid_hostnames`],
+    /// [`rustls::ClientConfig::dangerous`]
+    pub fn danger_no_tls_verify(&mut self, no_tls_verify: bool) -> &mut Self {
+        self.no_tls_verify = no_tls_verify;
+        self
+    }
+
+    /// Make a [`Client`] using the configuration.
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 993)
+    ///     .starttls().connect()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn connect(&self) -> Result<Client<ConnType>> {
+        let tcp = TcpStream::connect((self.domain.as_ref(), self.port))?;
+        #[cfg(any(feature = "rustls-tls", feature = "native-tls"))]
```

Comment:
I think the body of this function would look a lot nicer if structured as:
```rust
match self.mode {
```
with each arm that is only available with TLS carrying a `#[cfg]`.